### ISSUE TITLE
Allow LRScheduler to step at the end of batches

### ISSUE THREAD

**Issue by sotte** [GRAAL-Research/pytoune#15]
TL;DR: some LRSchedulers must be able to update the learning rate (and other parameters) during the batches and not just at the end of an epoch. Should we implement this? What is the best way?

*Long version:*

I would love to see PyToune offer advanced learning rate schedules like 1 cycle policy!

PyTorch's `_LRScheduler` have a `step()` method which set the learning rate. The user is then responsible to call the `step()` method somewhere in the training loop. This normally happens at the end of each epoch, but it could also happen after each batch.

PyToune wraps PyTorch's `_LRScheduler`s via the `Callback` interface and calls the `step()` method `on_epoch_end`. This is great for convenience, but some LRSchedulers have to update the lr after each batch.

Here are a few options I see:
- We could update the `PyTorchLRSchedulerWrapper` and make it configurable when to trigger the `step()` call, and then use it something like this:
  ```python
  CosineAnnealingLR(step_on_batch_end=True, ...)
  ```

- We could introduce batch versions of all LRSchedulers.

- I use a simple callback `LRSetter` that gets a list of learning rates (and momentum values) and sets the lr (and momentum) on each batch end.

**Comment by freud14** [comment]
I agree with the `step_on_batch_end` argument in `PyTorchLRSchedulerWrapper` which would call `step()` in `on_batch_end` instead of `on_epoch_end`. However, for other hperparameters, I don't think it is possible to do it with PyTorch's LR schedulers. Maybe a pull request could be made to PyTorch LR schedulers to support other hperparameters?

**Comment by sotte** [comment]
Extending PyTorch LRScheduler inteface would be quite a bit of work and they are focusing on the "production-readyness" aspect for the 1.0 release. But I guess one should give it a try.

Just to give you an idea how the interface I'm using looks like currently:
```python
class OptimizerPolicy(Callback):
    def __init__(self, name, values):
        self.name = name
        self.values = values
        # the model does not track the total number of steps taken
        # so we track it here
        self.current_step = 0
        
    def on_batch_begin(self, batch, logs):
        value = self.values[self.current_step]
        for group in self.model.optimizer.param_groups:
            group[self.name] = value
        self.current_step += 1

lr_policy = OptimizerPolicy("lr", Cosine(eta_max=0.1, eta_min=0.001, T_max=len(train_dl)))                            
mom_policy = OptimizerPolicy("momentum", Cosine(eta_max=0.0, eta_min=0.06, T_max=len(train_dl)))
# mom_policy = OptimizerPolicy("momentum", Linear(T_max=len(train_dl), eta_max=0.0, eta_min=0.06))

model.fit_generator(
    train_dl,
    val_dl,
    epochs=10,
    callbacks=[
        lr_policy,
        mom_policy,
    ],
)
```
(That being said, I'm experimenting with different ways to specify the training phases. The folks from fastai have a nice interface: https://github.com/sgugger/Deep-Learning/blob/master/Understanding%20the%20new%20fastai%20API%20for%20scheduling%20training.ipynb)

**Comment by sotte** [comment]
In that case I'll just keep you posted on my "interface experiments" and we can figure out an API together.

**Comment by sotte** [comment]
Having different `OptimizerPolicy`s for different aspect of the optimizer (e.g. one for the lr, one for the momentum) is nice, as long as you don't want to specify different training phases.
```python
steps = len(train_dl)

# this look nice
lr_policy = OptimizerPolicy( "lr", [
    np.linspace(0.001, 0.1, steps),
    Cosine(eta_max=0.1, eta_min=0.001, T_max=steps),
    Cosine(eta_max=0.1, eta_min=0.001, T_max=steps*2),
])
# but it gets cumbersome if you want to set multiple parameters.
# now you have to make sure that the length of momentum matches the lr
mom_policy = OptimizerPolicy("momentum", [
    np.linspace(0, 0, steps),
    Cosine(eta_max=0.0, eta_min=0.06, T_max=3*steps),
])
    
model.fit_generator(train_dl, val_dl, epochs=100,
                    callbacks=[lr_policy, mom_policy]);
```

The more I play with the *fast.ai training phase interface* the more I like it. Here is a slightly adjusted version:
```python
steps = len(train_dl)

phases = [
    # If we set the optimizer, we would have to overwrite the default one.
    # Do we want that? We could just leave out this feature.
    Phase(steps=steps, opt_fn=SGD, lr=0.01),
    Phase(steps=3*steps, lr=0.01),  # fix value
    Phase(steps=3*steps, lr=(0.01, 0.1)),  # linear by default
    Phase(steps=3*steps, lr=(0.01, 0.1, "linear")),  # explicet
    Phase(steps=3*steps, lr=(0.01, 0.1, "cosine")),
    Phase(steps=2*steps, lr=(0.1, 0.001, "linear"),
                         momentum=(0.1, 0.2)),
]
# Building cosine by hand
for i in range(3):
    phases.append(Phase(steps=steps*(cycle_mult ** i), lr=(0.001, 0.1, "cosine")))

# Pre-build cosine phase
phases += CosinePhase(cycle_len=steps, lr=(0.001, 0.1), cycle_mult=2)
phases += OneCycle(cycle_len=steps*42, lr=(0.001, 0.1), momentum=(0.95, 0.85))

model.fit_generator(train_dl, val_dl, epochs=100,
                    callbacks=[OptPolicy(phases)])
```
A phase is just an iterator that returns a dict of the values to set `{"lr": 0.1, "momentum":  0.9}`.

**Comment by sotte** [comment]
Here is a notebook with some experiments. The fast.ai interface is quite nice.
https://github.com/sotte/pytoune/blob/optimizer_interface_experiments/examples/optimizer_policy_experiments.ipynb

There are still some open issues:
- There should be a way to specify that phases should repeat x times or even indefinitely, e.g. the SGDR policy with multiplier for the cosine policy.
- There should be a way to end the training from the Phase (e.g. after the OneCycle policy). This is easy to implement, but I'm not so sure about the interface.
- Do we want to be able to set the optimizer via the Phase?

Some more thoughts:
```python
phases = [
    # we could also pass the values in directly and specify how often to repeat them
    Phase(repeat=2, lr=np.linspace(1, 0.1, 100),
    # the values could also be generated by an infinite generator, here Cosine
    Phase(repeat=-1, lr=Cosine(1, 0.1, 100, cycle_mult=2)),
]
```

**Comment by freud14** [comment]
Hi, thank you for all the details. I like what I see. I am not sure I understand all of the details, but It trust you that you will come to a nice interface. 

It sounds like there are a lot of features you want to integrate. Maybe you should try to implement a simpler solution, test it thoroughly with unit tests, document everything, and afterwards add features incrementally? Unless you think you have all the time to implement everything? I would be happy in either case. Of course, every feature you think of should still be planned to be implementable using your interface.

Anyway, I like the `Phase` interface. I also like that you thought about the pre-built phase classes. I was going to suggest that before I saw the examples.

> There should be a way to end the training from the Phase (e.g. after the OneCycle policy). This is easy to implement, but I'm not so sure about the interface.

I'm not sure what you mean here. Could not you just set the number of epoch for the end of your policy?

> Do we want to be able to set the optimizer via the Phase?

It could be done. However, I would require the user to pass either an already initialized optimizer so that the user can decide which parameters to optimize and pass other initialization arguments.

Also, like the LR schedulers, it could be nice to be able to save an optimizer policy. Maybe we only need to save the current step in the `OptimizerPolicy` class?

**Comment by sotte** [comment]
Hey @freud14, 

thanks for the feedback! Ya, there are lot's of parts that have to be implemented. Starting small and adding stuff is the right way to go. I just wanted to get an idea of where I want to go before that. I think I'm done with my experiments now.

This is the current state. (Nothing is final here, especially the name are not :) )
- I have a `Phase` that implements `__getitem__` and `__len__` and returns a dict of optimizer parameters to set.
- `TrainPolicy` get a list of phase and iterates through them and sets the optimizer parameters.
- Complex policies such as SGDR and OneCycle are just a list of phases: less code and pretty simple. Alternatively one could implement a sep. `Phase`.

What I don't like:
- The parameter spaces ("lin", "cosine", etc.) create a list/ndarray. That can take up some memory. Think 1000 epochs and with 1000 steps each. We could and should just lazily calculate the values in `__getitem__` of `Phase`. Or maybe write little classes for that.
- `TrainPolicy` does not save the state and can't be saved and restarted. That feature should be added.

[Here](https://github.com/sotte/pytoune/blob/optimizer_interface_experiments/examples/train_policy_experiments.ipynb) is a notebook showing the current interface. And [here](https://github.com/sotte/pytoune/blob/optimizer_interface_experiments/pytoune/framework/policies.py) is a quick and dirty implementation.

If it's ok with you I would create a pull request, specify the basic features, add the implementation and test it. From there we can add more features.

**Comment by freud14** [comment]
Alright, I'm fine with what I see.

> `TrainPolicy` does not save the state and can't be saved and restarted. That feature should be added.

Maybe just an `initial_step` argument that would start the `TrainPolicy` at some step which is not at the start. Of course, this will modify your implementation a bit.

Also, it would be nice to support infinite phases which would simply not implement the `__len__()` method.

I'm waiting for your PR.
